base/show.jl

```diff
@@ -1,9 +1,60 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
 show(x) = show(STDOUT::IO, x)
-
 print(io::IO, s::Symbol) = (write(io,s);nothing)
 
+immutable IOContext{IO_t <: IO} <: AbstractPipe
+    io::IO_t
+    lock::ReentrantLock
+    limit_output::Bool
+    hasproperty::Bool
+    propertykey::Any
+    propertyvalue::Any
+    parent::IOContext
+    function IOContext(io::IO, limit_output::Bool=false)
+        assert(!isa(io, IOContext))
+        new(io, ReentrantLock(), limit_output, false, nothing, nothing, #=null=#)
+    end
+    function IOContext(io::IO, KV::Pair)
+        assert(!isa(io, IOContext))
+        new(io, ReentrantLock(), false, true, first(KV), last(KV), #=null=#)
+    end
+    IOContext(io::IOContext, limit_output::Bool) =
+        new(io.io, io.lock, limit_output, false, nothing, nothing, io)
+    IOContext(io::IOContext, KV::Pair) =
+        new(io.io, io.lock, io.limit_output, true, first(KV), last(KV), io)
+end
+IOContext(io::IO, arg=false) = IOContext{typeof(io)}(io, arg)
+IOContext(io::IOContext, arg=io.limit_output) = typeof(io)(io, arg)
+IOContext(io::IOContext, key, value) = typeof(io)(io, Pair{Any, Any}(key, value))
+show(io::IO, ctx::IOContext) = (print(io, "IOContext("); show(io, ctx.io); print(io, ")"))
+
+pipe_reader(io::IOContext) = io.io
+pipe_writer(io::IOContext) = io.io
+lock(io::IOContext) = (lock(io.io); lock(io.lock)) # double lock, because IO itself might not be lockable
```

Comment:
there isn't an extra layer. it is unknown if the IO object that the IOContext is wrapping has a lock, so IOContext brings it own to ensure that there is at least one lock